_posts/2019-12-31-memkind-dax-kmem.md

```diff
@@ -0,0 +1,215 @@
+node 1 size: 80608 MB
+node 1 free: 71958 MB
+node 2 cpus:
+node 2 size: 1026048 MB
+node 2 free: 1026048 MB
+node 3 cpus:
+node 3 size: 512000 MB
+node 3 free: 512000 MB
+node distances:
+node   0   1   2   3
+  0:  10  21  17  28
+  1:  21  10  28  17
+  2:  17  28  10  28
+  3:  28  17  28  10
+{% endhighlight %}
+
+The following figure corresponds to it:
+
+![memkind_dax_kmem](/assets/memkind_dax_kmem_config.png)
+
+Note:
+- Image is simplified to present only 2 CPUs: CPU 0 which belong to Node 0 and CPU 30 which belong to Node 1
+- Numbers in squares corresponds to NUMA node distances between Nodes 0/1 and all of the Nodes
+- Node 0 and Node 1 are based on DRAM memory where Node 2 and Node 3 are based on NVDIMM memory
+- Moving physical memory to swap space is limited only to one Node
+- To simplify examples below we assume that process uses only CPU 0
+
+**MEMKIND_DAX_KMEM**:
+This is the first variant where the allocation comes from the closest persistent memory NUMA node
+at the time of allocating.
+
+![memkind_dax_kmem](/assets/memkind_dax_kmem.gif)
+
+The process runs only CPU 0, CPU which is assigned to Node 0. Node 2 is the closest persistent memory NUMA node to Node 0,
+therefore using **MEMKIND_DAX_KMEM** results in taking memory only from Node 2. If there is not enough free memory in Node 2
+to satisfy an allocation request, inactive pages from Node 2 are moved into the swap space - freeing up the memory for
+**MEMKIND_DAX_KMEM** use case.
+
+**MEMKIND_DAX_KMEM_ALL**:
+This is the second variant where the allocation comes from the closest persistent memory NUMA node
+available at the time allocating.
+
+![memkind_dax_kmem_all](/assets/memkind_dax_kmem_all.gif)
+
+A similar situation to scenario presented in **MEMKIND_DAX_KMEM** except that when there is not enough free
+memory in Node 2 to satisfy the allocation request, an allocation pattern switch to Node 3. When available free space is
+exhausted in Node 3 - swap space is used.
+
+**MEMKIND_DAX_KMEM_PREFERRED**:
+This is the third variant where the allocation comes from the closest persistent memory NUMA node at the time
+of allocating. If there is not enough memory to satisfy the request the allocation will fall back on other memory
+NUMA nodes.
+
+![memkind_dax_kmem_preferred](/assets/memkind_dax_kmem_preferred.gif)
+
+Again the allocation starts from Node 2. When there is not enough free memory in Node 2, the allocation switches to other
+nodes in order of increasing distance from the preferred node based on information provided by the platform firmware:
+Node 0, Node 3 and ending in Node 1. When available free space is
+exhausted in Node 1 - swap space is used.
+
+
+**Caveats**
+
+For **MEMKIND_DAX_KMEM_PREFERRED**, the allocation will not succeed if two or more
+persistent memory NUMA nodes are in the same shortest distance to the same CPU on which the process is eligible to run.
```

Comment:
Eligible is an odd word to use here. "Permitted", "Allowed" ?

Reply:
Done - just FYI i take this eligible from sched_setaffinity(2) manual page

Reply:
If it's from the man page then leave it as eligible. It still sounds odd to me :P